### PR TITLE
Fixes #119

### LINKED ISSUE #119: _validate_suffix_collections needs rework
I have recently hit an edge case where the name of my resource had to contain '/' this has led to problems with our validation method: 

`````

C:\Users\wypior\Documents\PythonProjects\f5-common-python\f5\bigip\tm\gtm\test\functional\test_topology.py:57: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
C:\Users\wypior\Documents\PythonProjects\f5-common-python\f5\bigip\tm\gtm\test\functional\test_topology.py:47: in delete_topology
    name=name)
C:\Users\wypior\Documents\PythonProjects\f5-common-python\f5\bigip\tm\gtm\topology.py:145: in load
    return self._load(**kwargs)
C:\Users\wypior\Documents\PythonProjects\f5-common-python\f5\bigip\resource.py:919: in _load
    response = refresh_session.get(base_uri, **kwargs)
C:\Python27\Lib\site-packages\icontrol\session.py:221: in wrapper
    sub_path, suffix, **kwargs)
C:\Python27\Lib\site-packages\icontrol\session.py:177: in generate_bigip_uri
    _validate_uri_parts(base_uri, partition, name, sub_path, suffix)
C:\Python27\Lib\site-packages\icontrol\session.py:151: in _validate_uri_parts
    _validate_name_partition_subpath(name)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

element = 'ldns: subnet 192.168.100.0/24 server: subnet 192.168.200.0/24'

    def _validate_name_partition_subpath(element):
        # '/' and '~' are illegal characters
        if element == '':
            return True
        if '~' in element:
            error_message =\
                "instance names and partitions cannot contain '~', but it's: %s"\
                % element
            raise InvalidInstanceNameOrFolder(error_message)
        elif '/' in element:
            error_message =\
                "instance names and partitions cannot contain '/', but it's: %s"\
                % element
>           raise InvalidInstanceNameOrFolder(error_message)
E           InvalidInstanceNameOrFolder: instance names and partitions cannot contain '/', but it's: ldns: subnet 192.168.100.0/24 server: subnet 192.168.200.0/24

C:\Python27\Lib\site-packages\icontrol\session.py:123: InvalidInstanceNameOrFolder

`````

This issue is blocking the implementation of GTM endpoints in our SDK

### PR DESCRIPTION
**Pull request by wojtek0806**
Problem:

_validate_suffix_collections method assumes that '\~' and '/' in elements are incorrect. This leads to issues where 'name' contains '~' or '/' on purpose.

Analysis:
This patch introduces name_transform bool (with default being False), with it set to True, the 'name' element will not be validated for the presence of '\~' and '/' characters. It will also convert all '/' in the name into '~'. To use this ability BIGIP needs to pass the argument 'name_transform' in the **kwargs when calling the iControl.

Tests:
unit

### CODE REVIEW

**Comment by wojtek0806** [comment]
@mattgreene @swormke @pjbreaux @zancas

File: icontrol/test/unit/test_session.py [hunk 1]
```diff
@@ -52,6 +63,17 @@ def uparts_with_subpath():
     return parts_dict
 
 
+@pytest.fixture()
+def transform_name_w_subpath():
+    parts_dict = {'base_uri': 'https://0.0.0.0/mgmt/tm/root/RESTiface/',
+                  'partition': 'BIGCUSTOMER',
+                  'name': 'foobar1: 1.1.1.1/24 bar1: /Common/DC1',
```
Comment (pjbreaux) [inline icontrol/test/unit/test_session.py:75]: Is this a valid name on the BIG-IP?

**Comment by pjbreaux** [comment]
I'd like to see a set of functional tests as well with this. Perhaps creating a partition/folder with a name that has slashes? Things like that.

File: icontrol/test/unit/test_session.py (hunk 1 above)
Comment (wojtek0806) [reply icontrol/test/unit/test_session.py:75]: yep, as discussed this is a valid name for gtm topology records

**Comment by wojtek0806** [comment]
ill create and try to load the record as discussed on slack

**Comment by wojtek0806** [comment]
@pjbreaux added functional tests feel free to test on GTM enabled 11.5.4 unit and merge when ready :)

**Comment by pjbreaux** [comment]
@wojtek0806: Could you post your test results here. Then I can merge.

Thanks

**Comment by wojtek0806** [comment]
https://gist.github.com/wojtek0806/db45d30ab419214348ab753950596f38
https://gist.github.com/wojtek0806/ec5047b75bc9c57054011c435d1b98c9

@pjbreaux

**Comment by pjbreaux** [comment]
We'll need to paste the results of all the tests, ensuring no regressions occurred.

**Comment by wojtek0806** [comment]
https://gist.github.com/wojtek0806/5f41b0b695e9084b56b7e77a62139112

The tests for auth fail cause i am running these against 11.5.4, those are unrelated to the changes I have 
made

@pjbreaux

**Comment by pjbreaux** [comment]
I don't see all of the tests being run here, just the auth failures. I think we still should test against a version where the auth tests succeed, just to ensure we haven't made any changes that haven't broken auth. And to get another data point of testing against a new device version.

**Comment by wojtek0806** [comment]
i am sorry but if you want to test this test it, please provide a jerkins like solution here, or you can pull my fork run the tests yourself, also my tests are aimed at 11.5.4 and 11.6.0 specifically due to how buggy GTM topology is.

**Comment by pjbreaux** [comment]
Running it against 11.5.4 and 11.6.0 sounds fine. For now it is manual, but I agree, it must be a Jenkins job or part of a build.

**Comment by wojtek0806** [comment]
@pjbreaux I have asked @caphrim007 to attach this to Jenkins, then I can rework the entire testing suite we have so far for functional tests and we can run this cross versions and do what we do in f5-sdk. I hate manual tasks with passion :)

**Comment by wojtek0806** [comment]
See attached tests, those are exports from Pycharm,

11.5.4 Fails Auth tests due to  BIGIP sending us 404 response, could be because technically tokens were not implemented in 11.5.x but this is just a guess. 11.6.0 tests pass "test_get_special_name_11_x_12_0" and "test_get_special_name_12_1" since it handles the extra spaces in resource names.

Tests test_get_special_name_11_x_12_0 pass in all versions but 12.1.0
Version 12.0.0 has a bug that prevents us from creating objects in the first place.



[Test Results - all.zip](https://github.com/F5Networks/f5-icontrol-rest-python/files/636092/Test.Results.-.all.zip)


@pjbreaux

**Comment by pjbreaux** [comment]
Could you skip the appropriate tests and just mark up with a message as to why the auth tests fail against 11.5.4 and why the special name test fails in 12.1.

**Comment by pjbreaux** [comment]
After that, I can merge it in.

**Comment by wojtek0806** [comment]
hmm i am not sure how do i do that, i only run pytest on pycharm, so what exactly do you wanna do? just pytest.skip ?

**Comment by pjbreaux** [comment]
Yes, just the skips and a message along with it.

**Comment by wojtek0806** [comment]
see the tests:

https://gist.github.com/wojtek0806/41e8af87dd424508a3f3141dee73e1ab#file-11-5-4test
https://gist.github.com/wojtek0806/fc70a5d39ee5ee32f34e97d1b7149504#file-1160_test
https://gist.github.com/wojtek0806/6ecaaa3a04a56a650fb9cf71f5818933#file-12-1-0_test

@pjbreaux @zancas
